examples/swapi/swapi-server.js

```diff
@@ -114,13 +121,60 @@ const createSWAPIServer = () => {
         }
     }
 
+    class FilmModelV2 {
+        id: number;
+
+        constructor(id: number) {
+            this.id = id;
+        }
+
+        async title() {
+            const response = await swapiLoaders.getFilmsV2.load({ film_id: this.id, property: 'title' });
+
```

Comment:
can you show the output of `$ flow check` when you try to read `response.director`?

hopefully it errors, which would verify the data masking is correct

Reply:
I added `console.log(response)` there, and it only gives me `film_id` and `title`. I could also do the flow check.

Reply:
I added `console.log(response)` and `console.log(response.director)`. 
got this 
```
{ film_id: 4, title: 'The Phantom Menace' }
undefined
```
but flow doesn't complain, which is something I should fix.

Reply:
I checked the response type just matches the SWAPI_Film_v2 type I defined [here](https://github.com/Yelp/dataloader-codegen/pull/261/files#diff-e461ae547272d09d3b85a44a7ab8de37f10cc4a08ae23a9ffb78b247b4425b7eR66-R73)
which is 
```
export type SWAPI_Film_v2 = $ReadOnly<{|
    film_id: number,
    title: string,
    episode_id: number,
    director: string,
    producer: string,
|}>;
```
so it won't blow up when I do `response.director`.

Reply:
Thanks!

I think we want flow to blow up here though don’t we, otherwise there’s a mismatch between what actually exists vs what flow thinks exists?

Reply:
@magicmark flow blow up here would be nicer! I've looked into how flow type is set for the response, seems they're these lines: https://github.com/Yelp/dataloader-codegen/blob/0aaa906ea0e31c7cd03dd613cb4e4c74dacde135/src/genTypeFlow.ts#L88-L92 

I don't fully understand the code ^, but I guess it's to extract the return type from the [function](https://github.com/Yelp/dataloader-codegen/pull/261/files#diff-e461ae547272d09d3b85a44a7ab8de37f10cc4a08ae23a9ffb78b247b4425b7eR93-R96)

Do you think it's feasible to change the code so that return type only contains the property we ask for?

Reply:
update: 
I don't think it's necessary to make flow blow up here. 

take `getFilmsV2` as example,
```
    getFilmsV2: ({|
        film_ids: $ReadOnlyArray<number>,
        properties: $ReadOnlyArray<string>,
    |}) => Promise<$ReadOnlyArray<SWAPI_Film_v2>>,
```
**status quo (without property batching):**
const response = await swapiLoaders.getFilmsV2.load({ film_id: this.id, **properties: ['title']** });
the flow type of response will just be 
```
export type SWAPI_Film_v2 = $ReadOnly<{|
    film_id: number,
    title: string,
    episode_id: number,
    director: string,
    producer: string,
|}>;
```
NOTE: we only asked for `title` here, but the response type would just be the SWAPI_Film_v2. Flow won't throw error when we do  `response.episode_id` or `response.director`, etc. 
So I think it's okay to keep the return type same as the status quo as long as the actual return value is correct.

Reply:
> status quo (without property batching):
>
> const response = await swapiLoaders.getFilmsV2.load({ film_id: this.id, properties: ['title'] });
> the flow type of response will also be SWAPI_Film_v2
> 
>  export type SWAPI_Film_v2 = $ReadOnly<{|
>     film_id: number,
>    title: string,
>    episode_id: number,
>    director: string,
>    producer: string,
> |}>;

this is wrong? it may be the "status quo", but this is just how the programmer defined the SWAPI_Film_v2 type. It doesn't take into account that some fields may or may not be present (indicated by nullable properties). (Just because something is status quo doesn't mean it's correct, it just means at least there's no regression)

We _know_ that if you didn't ask for `director`, you wont get a `response.director` property. That's the contract.

At a very minumum, we should mark all properties as nullable - the more correct thing to do would be to omit all non-requested properties from the response entirely.

Reply:
`At a very minumum, we should mark all properties as nullable` I think we get this for free. 
The swagger endpoint defines only `id` is required in the response object. 

I think `the more correct thing to do would be to omit all non-requested properties from the response entirely.` is something not easy to achieve, especially we support three user cases here, and we need to adjust the flow type accordingly. I personally think it's a nice to have.

Reply:
> I think we get this for free. The swagger endpoint defines only id is required in the response object.

Hmm, here, we only get nullable _because_ the swagger endpoint defines each as property so. That's just what the example endpoint does - there's no guarantee other endpoints will do that.

I realize this is super fiddly :(

> not easy to achieve, especially we support three user cases

I think doing this 'dynamically' (i.e. at runtime) to sniff what shape the response is makes this harder (if not impossible?) to generate the correct flow types - hence the suggestion to encode this in the config file and declare it up front.

curious - what does this look like if we just start off supporting one use case?

Reply:
> I personally think it's a nice to have.

Similar to https://github.com/Yelp/dataloader-codegen/pull/261#discussion_r631284573:

> this layer is the funnel through which the majority of our data flows through, with lots of devs interfacing with this API and lots of opportunity for misuse :(

This is pretty low level infra - given how many callsites there are to dataloaders, I think correct flow types are a must here.

Reply:
@magicmark I have trouble with generate correct flow type. Here's a [try flow reproduce ](https://flow.org/try/#0PQKgBAAgZgNg9gdzCYAoAxnAdgZwC5gBiccAwjAIY45gC8YA3qmC2FCYQK5boAU2AUwBcYLJwC2AIwEAnADRgADgDkK44fhkBLLAHMwASkbNWpzLjgwBAOni7eKtQIMmWAX1QfUeAJ6KBYADKAOoAggAKAJIA+oRaMOLRAGoALHRgACQASgIUACYA8lgwPgA8DAA+rmzxiVp5ImJSsnLVeFp4ViKaOrqtpgKKWjhweQLR9Y0S0vLVeVoyAuh4cDLdeNp6-ayKMqOc6LLrm32oFW4AfADcqKgCAB6KqwS+-kQkACp+AfSV1exwLg8XjZXKFYplSqCKbNeRKVTqY69c4XAwicJ7cTDASlUH5IolUIyGQUMohCIxOIJZIpC4XVrnW4PJ4yF7fMChGAwL7+Gi-arEMiUagiQU8gSeW6vAIAUXuGwoyyJ+nopWVChyeAuvAA2rxlUZaBcwJqALqG43K7zsuUK5YYuBYnACPKazgyLBJCgwTg40LG1VZbW8ayh0IWsAOp04oMRrI3VDAYAPO0vAAWAUWeHdWDA0rAcCgbA43HQCgQaa06DTYGGkcx2NxOXxEKJJLJYSisVqNLp1reWZz4roqAypG9MFK1VtJPtDedroE2Y9Xp9OJ1Y4npRnirw6p11RYGQd-lZPnFU9MpmPe1PvgvnO53xwCgARILyFQcK-6YfWK+ASBdBXz-C5TTA39rxPWR72+S8rxvOA73PODH3FF8wHfEhP2oH9tlMACSx4EDTF-C5biTMA0zwPBFBwIQk1gRBrFWXRkywYA8jgdAcGAaVeM4doYA6LQBF4gBiFZ0AAWhwNMKH8RNgDASICwhGosDyPNKxoDJAnkt48ErPQwHQeBnTzOAwArCgCAQAIEAoLA8AUJSwEkQTtLrbASms1YAGsaHYGQC0kAArPNvgULA4AIJywAoYlSTAftMyXHNAk4SRnTwYd6D0gycXJLsqUSVJrlucx8DAQcPQARhEYrKR7VJ0iYFgoB7SYwDq7Z2k6YQwAAcgBIbtkGYZRnGbq6oAJgAZm2eZFmWVYRBGkgxpMXZ9kONZhtGzwE0onyfFEAQXUsmr0o9BKwBwLKcoLIsjICTqBBgPIcAwbBqtqrBZpEf7Muypc8uMDquoaHqloWJYVn2ja4C2lgdryA4jgOzajoo5S00QK7xAofy3qhq7pFEWLRE4LkKEkKwEs07S0uqlZ3ICMRafpgQAH4Uqqgh-vmoGbqwcGdXampqRmvqOi6LHkdhlaEfW0btjRjHEcOtxTRuIA)
Could you help me with that? I don't know how to make `id` not nullable while change other fields nullable. 

On the other hand, 
> Hmm, here, we only get nullable because the swagger endpoint defines each as property so. That's just what the example endpoint does - there's no guarantee other endpoints will do that.

I think we can document supported resource must return nullable properties? If some endpoint require all properties be non-nullable, and we remove lots of properties in the dataloader-codegen, that seems wrong to me and we shouldn't support that case. So we require properties being nullable make sense to me.

Reply:
there's a (subtle) difference between getting the superset of all possible null properties and the subset of known guaranteed non-null properties

So if I did:

```js
const properties1 = loaders.fetchStuff({ properties: ['foo', 'bar'] });
const properties2 = loaders.fetchStuff({ properties: ['baz'] });
```

If I understand what you're saying correctly, you'd expect both `properties1` and `properties2` to have the following identical type:

```jsx
{ foo?: string, bar?: string, baz?: string }
```

If we're being super nitpicky, this is wrong:
- there's no way that `properties1` should ever have `baz` present since we didn't request it. Flow is kinda lying to us.
- In`properties2`, `baz` should _always_ be present. Again, flow is kinda lying to us.

These aren't catastrophic problems. The stock price of the company won't go up or down as a result either way here :)

But again, as this is foundational infrastructure that hundreds of our resolvers use, I think we probably want to get the flow typings as "correct" as possible.

Reply:
You commented this before 

> At a very minumum, we should mark all properties as nullable - the more correct thing to do would be to omit all non-requested properties from the response entirely. 

So I thought just having all properties nullable is fine. 
I have trouble with getting it as "correct" as possible.

Reply:
>  we should mark all properties as nullable

Is that what the current PR does? Or does it rely on the endpoint already doing that for us? That's a key difference.

> I think we can document supported resource must return nullable properties? 

Is a reasonable alternative.

I'm happy to set up some time to pair on this to see if we can get flow to do the right thing? We can timebox it, and if not, fall back to your suggestion :)

Reply:
> Is that what the current PR does? Or does it rely on the endpoint already doing that for us? That's a key difference.

We rely on the endpoint returns nullable properties. And that's the reasonable alternative thing you commented on. 

> I'm happy to set up some time to pair on this to see if we can get flow to do the right thing? We can timebox it, and if not, fall back to your suggestion :)
Thanks!